graphql_compiler/schema_transformation/rename_schema.py

```diff
@@ -186,25 +217,27 @@ class RenameSchemaTypesVisitor(Visitor):
         }
     )
 
-    def __init__(self, renamings, query_type, scalar_types):
+    def __init__(
+        self, renamings: Dict[str, Optional[str]], query_type: str, scalar_types: Set[str]
+    ) -> None:
         """Create a visitor for renaming types in a schema AST.
 
         Args:
-            renamings: Dict[str, str], mapping from original type name to renamed type name.
-                       Any name not in the dict will be unchanged
+            renamings: Dict[str, Optional[str]], mapping from original type name to renamed type
+                       name. Any name not in the dict will be unchanged
             query_type: str, name of the query type (e.g. RootSchemaQuery), which will not
                         be renamed
             scalar_types: Set[str], set of names of all scalars used in the schema, including
                           all user defined scalars and any builtin scalars that were used
         """
         self.renamings = renamings
-        self.reverse_name_map = {}  # Dict[str, str], from renamed type name to original type name
-        # reverse_name_map contains all types, including those that were unchanged
+        self.reverse_name_map: Dict[str, str] = {}  # From renamed type name to original type name
+        # reverse_name_map contains all non-suppressed types, including those that were unchanged
         self.query_type = query_type
         self.scalar_types = frozenset(scalar_types)
         self.builtin_types = frozenset({"String", "Int", "Float", "Boolean", "ID"})
 
-    def _rename_name_and_add_to_record(self, node):
+    def _rename_name_and_add_to_record(self, node: Node) -> Union[ellipsis, Optional[Node]]:
```

Comment:
Stuck on type-hinting this function signature, specifically the return type.

This function can return a `Node`, `None`, or `REMOVE`. It turns out the GraphQL library [sets](https://github.com/graphql-python/graphql-core/blob/master/src/graphql/language/visitor.py#L34) `REMOVE` to the Python literal `Ellipsis`, not `None`-- I think PyCharm got confused because there's a stub file that sets `Ellipsis` to `None`, but when this code actually runs `Ellipsis != None`.

The current error says that this type-hint isn't valid and `note: Suggestion: use type[...] instead of type(...)`.

* type[ellipsis] and type[type(Ellipsis)] don't work because of the square brackets.

* Although `type(Ellipsis)` is the class `ellipsis`, adding `ellipsis` to the union in place of `type(REMOVE)` makes both tests and linter to fail because `ellipsis` is undefined.

* Replacing `type(REMOVE)` with `Type[type(REMOVE)]` also makes the linter fail and produces a link to [this documentation](https://mypy.readthedocs.io/en/latest/common_issues.html#variables-vs-type-aliases). It seems unrelated to what the type hint we're trying to do here because I a) haven't been able to find the `ellipsis` class and b) it makes more sense to depend on the type of `REMOVE` as much as possible rather than rely on the fact that it's set to `Ellipsis`.

Reply:
Try `Literal[Ellipsis]` instead? So `Union[Literal[Ellipsis], Optional[Node]]` for the return type hint.

Reply:
Nope :(

```
error: Parameter 1 of Literal[...] is invalid
error: Variable "builtins.Ellipsis" is not valid as a type
```

Reply:
I went through:
https://github.com/python/typing/issues/684
https://github.com/python/mypy/issues/2180
https://github.com/python/mypy/issues/7818

The consensus seems to be that using `Ellipsis` as an "easy-to-understand, easy-to-import sentinel value" is not recommended and will lead to type issues exactly like we're seeing here. The issues are closed as "won't fix", so I'd recommend checking out the latest version of GraphQL-core (I think we're a minor version behind or so), and if the use of `Ellipsis` is still present, opening a GitHub issue. You could link back to this PR (ideally to this comment thread, by referencing the new issue from here), and ask the maintainers there to create a separate singleton sentinel value per the BDFL's guidance in the GitHub issues I listed above.

In the meantime, I'd recommend making a type alias like `VisitorReturnType = Union[Optional[Node], Any]`, documenting and explaining the necessity of using `Any`, linking to the GraphQL-core issue you open, and using that type alias everywhere it's needed.

Reply:
Done! One note: In a separate PR, I'll be updating all the existing `return None`s to `return IDLE` where it makes sense since the GraphQL library defines that for doing nothing. See [here](https://github.com/kensho-technologies/graphql-compiler/pull/834#discussion_r436889940) explaining why the `VisitorReturnType` won't have the `Optional` part.